### PR TITLE
Add exchange state machine

### PR DESCRIPTION
**Pull request by diehuxx**
This state machine class `Exchange` will be used to great effect in a follow-up PR in the http-server package.

### CODE REVIEW

**Review by mistermoe — APPROVED** [review]
amajing

File: .changeset/green-cheetahs-share.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
```
Comment (jiyoontbd) [inline .changeset/green-cheetahs-share.md:1]: do they tho? i feel like cheetahs aren't good at sharing.

**Review by jiyoontbd — APPROVED** [review]

File: .changeset/green-cheetahs-share.md (hunk 1 above)
Comment (diehuxx) [reply .changeset/green-cheetahs-share.md:1]: green cheetahs are envious cheetahs

File: packages/protocol/tests/exchange.spec.ts [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+        for (const message of [rfq, order, orderStatus]) {
+          try {
+            exchange.addNextMessage(message)
+            expect.fail()
+          } catch (e) {
+            expect(e.message).to.contain('is not a valid next message')
+          }
+        }
+      })
+
```
[the remark this answers is not in the document]
Comment (diehuxx) [reply packages/protocol/tests/exchange.spec.ts:201]: Good call. Done.

**Pull request merged**